src/java.base/share/classes/java/util/UUID.java

```diff
@@ -99,24 +104,190 @@ public final class UUID implements java.io.Serializable, Comparable<UUID> {
+            //
+            // Without recording the buffer index in the thread itself, the good option is to use
+            // the thread ID scrambled with Murmur hash, which results in good bit entropy.
+
+            long h = RandomSupport.mixMurmur64(Thread.currentThread().threadId());
+            int idx = (int)(h & (BUFS_COUNT - 1));
+            Buffer current = BUFS[idx];
+            if (current == null) {
+                // Create a new buffer and install it.
+                // On initial contention, some buffers may be lost, but this is not
+                // a problem for correctness, or for steady-state performance.
+                current = new Buffer(newRandom());
+                BUFS[idx] = current;
+            }
+            return current.next();
+        }
+
+        // Buffer random reads. This allows batching the SecureRandom provider requests.
+        // Current implementation targets the 4K buffer size, which balances the initialization
+        // costs, memory footprint and cache pressure.
+        @jdk.internal.vm.annotation.Contended
+        static final class Buffer {
+            static final String PROP_NAME_UUID_COUNT = "java.util.UUID.uuidsPerBuffer";
+
+            static final int UUID_CHUNK = 16;
+            static final int UUID_COUNT;
+            static final int BUF_SIZE;
+
+            static final VarHandle VH_POS;
+            static {
+                try {
+                    VH_POS = MethodHandles.lookup().findVarHandle(Buffer.class, "pos", int.class);
+                    UUID_COUNT = Integer.getInteger(PROP_NAME_UUID_COUNT, 256);
+                    if (UUID_COUNT < 1 || UUID_COUNT > Integer.MAX_VALUE / UUID_CHUNK) {
+                        throw new IllegalArgumentException(PROP_NAME_UUID_COUNT + " is out of range");
+                    }
+                    BUF_SIZE = UUID_CHUNK * UUID_COUNT;
+                } catch (Exception e) {
+                    throw new InternalError(e);
+                }
+            }
+
+            final SecureRandom random;
+            final StampedLock lock;
+            final byte[] buf;
+            int pos;
+
+            public Buffer(SecureRandom random) {
+                this.random = random;
+                this.lock = new StampedLock();
+                this.buf = new byte[BUF_SIZE];
+                this.pos = BUF_SIZE; // trigger re-creation on first use
+            }
+
+            public UUID next() {
+                long stamp = lock.tryOptimisticRead();
+                try {
+                    UUID uuid = null;
+
+                    // Optimistic path: optimistic locking succeeded.
+                    // Try to pull the UUID from the current buffer at current position.
+                    if (stamp != 0) {
+                        int p = (int)VH_POS.getAndAdd(this, UUID_CHUNK);
+                        if (p < BUF_SIZE) {
+                            uuid = new UUID(buf, p);
```

Comment:
We can read msb/lsb from the buffer here and move object allocation outside the lock to reduce the length of the critical section.

Reply:
Yes, I am doing this.